simulator/A-NC/utils.py

```diff
@@ -0,0 +1,344 @@
+    The class inherits from the Module class of torch.nn.
+    """
+
+    def __init__(self):
+        """
+        Initializes the neural network.
+        """
+        super(Net, self).__init__()
+        self.conv1 = nn.Conv2d(1, 32, kernel_size=3, stride=2, padding=1)
+        self.conv2 = nn.Conv2d(32, 64, kernel_size=3, stride=2, padding=1)
+        self.conv3 = nn.Conv2d(64, 128, kernel_size=3, stride=2, padding=1)
+        self.fc1 = nn.Linear(2048, 128)
+        self.fc2 = nn.Linear(128, 10)
+
+    def forward(self, x):
+        """
+        Performs the forward pass of the neural network.
+
+        Params:
+            x (torch.Tensor): the input vector
+
+        Returns:
+            output (torch.Tensor): the output vector
+        """
+        x = self.conv1(x)
+        x = F.relu(x)
+        x = self.conv2(x)
+        x = F.relu(x)
+        x = self.conv3(x)
+        x = F.relu(x)
+        x = torch.flatten(x, 1)
+        x = self.fc1(x)
+        x = F.relu(x)
+        x = self.fc2(x)
+        output = F.log_softmax(x, dim=1)
+        return output
+
+
+class Net2(nn.Module):
+    """
+    A class defining a neural network with one convolutional layer and
+    one fully-connected layer. The activation function is ReLU except for
+    the final layer, where it is log_softmax.
+    The class inherits from the Module class of torch.nn.
+    """
+
+    def __init__(self):
+        """
+        Initializes the neural network.
+        """
+        super(Net2, self).__init__()
+        self.conv1 = nn.Conv2d(1, 32, kernel_size=3, stride=2, padding=1)
+        self.fc1 = nn.Linear(6272, 10)
+
+    def forward(self, x):
+        """
+        Performs the forward pass of the neural network.
+
+        Params:
+            x (torch.Tensor): the input vector
+
+        Returns:
+            output (torch.Tensor): the output vector
+        """
+        x = self.conv1(x)
+        x = F.relu(x)
+        x = torch.flatten(x, 1)
+        x = self.fc1(x)
+        output = F.log_softmax(x, dim=1)
+        return output
+
+def client_update(client_model, optimizer, train_loader, epoch=5):
+    """
+    Trains a client model on the training data using an optimizer for
+    some number of epochs.
+
+    Params:
+        client_model (nn.Module): the given neural network
+        optimizer (SGD): the given optimizer algorithm
+        train_loader (DataLoader): the training dataset
+        epoch (int): the number of times the optimizer is run on the entire dataset
+
+    Returns:
+        loss (float): the training loss of the model after the last iteration
+    """
+    client_model.train()
+    for e in range(epoch):
+        for batch_idx, (data, target) in enumerate(train_loader):
+            data, target = data.cuda(), target.cuda()
+            optimizer.zero_grad()
+            output = client_model(data)
+            loss = F.nll_loss(output, target)
+            loss.backward()
+            optimizer.step()
+    return loss.item()
```

Comment:
in addition to the loss, we may need to track other metrics, e.g. top-1 acc?
also, right now, it seems like the function will only return the loss value of the last iterate. I think it is not sufficient?>